### PR TITLE
Add documentation for Kinesis source in Data prepper

### PR DESCRIPTION
**Pull request by sb2k16**
### Description
This PR adds documentation for a new `kinesis` source in Data Prepper. It includes the following:
- Example configuration
- Required configuration attributes along with their description
- Required permissions to run `kinesis` source

### Issues Resolved
N/A

### Checklist
- [X] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by vagimeli** [comment]
@dlvenable This PR is tagged as a first-time contributor. Will you or your team provide a technical review? Once that review is done, I'll do a doc review. Thank you.

**Comment by vagimeli** [comment]
@sb2k16 Is this PR ready for a doc review? I can give it a first review and then give it a final review before the Data Prepper 2.10.

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+Option | Required | Type    | Description
+:--- | :--- |:--------| :---
+`skip_lines` | No | Integer | The number of lines to skip before creating events. You can use this configuration to skip common header rows. Default is `0`.
+`header_destination` | No | String  | A key value to assign to the header line of the stream event. If this option is specified, then each event will contain a `header_destination` field.
+
+### polling
+Option | Required | Type    | Description
+:--- | :--- |:--------| :---
+`max_polling_records` | No | Integer | The number of records to fetch from Kinesis during a single call to get Kinesis stream records.
+`idle_time_between_reads` | No | Integer  | The time duration to sleep in between calls to get Kinesis stream records.
```
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:77]: This is a `Duration`, not an `Integer`.

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+layout: default
+title: kinesis
+parent: Sources
+grand_parent: Pipelines
+nav_order: 3
+---
+
+# kinesis
+
+You can use `kinesis` source in Data Prepper to ingest records from one or more [Amazon Kinesis](https://aws.amazon.com/kinesis/) Data Streams.
```
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:11]: Should we link here?

https://aws.amazon.com/kinesis/data-streams/

And then update the anchor to surround all of "Amazon Kinesis Data Streams?"

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+`streams` | Yes      | List     | Multiple Kinesis Data streams that Data Prepper `kinesis` source uses to read records. You can configure up to 4 streams. For more information about `streams` configuration options, see [Streams](#streams).
+`codec` | Yes      | Codec    | The [codec](#codec) to apply.
+`buffer_timeout` | No       | Duration | The amount of time allowed for writing events to the Data Prepper buffer before timeout occurs. Any events that the source cannot write to the buffer during the specified amount of time are discarded. Default is `1s`.
+`records_to_accumulate` | No       | Integer  | The number of messages that accumulate before being written to the buffer. Default is `100`.
+`consumer_strategy` | No       | String   | Consumer strategy to use for ingesting Kinesis data streams. Default is `fan-out`. However, `polling` can also be used. if `polling` is enabled, additional configuration for `polling` will need to be added.
+`polling` | No       | String   | if `consumer_strategy` is set to `polling`, this config will need to be added. Refer to [polling](#polling).
```
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:45]: The type here is `polling`. Please update the type column and have it link to the `polling` section below.

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Comment by sb2k16** [comment]
> @sb2k16 , @vagimeli , This is a new feature for Data Prepper 2.10, which has not yet been released. We should target this to a new branch. I just created one named `data-prepper-2.10`.
> 
> @sb2k16 , Can you edit your PR to target the `data-prepper-2.10` branch?

I have edited the PR to target `data-prepper-2.10` branch.

File: _data-prepper/pipelines/configuration/sources/kinesis.md (hunk 1 above)
Comment (sb2k16) [reply _data-prepper/pipelines/configuration/sources/kinesis.md]: Thanks @dlvenable. I have made the changes.

File: _data-prepper/pipelines/configuration/sources/kinesis.md (hunk 2 above)
Comment (sb2k16) [reply _data-prepper/pipelines/configuration/sources/kinesis.md]: Thanks @dlvenable. I have made the changes.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -78,6 +78,8 @@ Follow these steps to set up your local copy of the repository:
 
 1. Navigate to your cloned repository.
 
+##### Building using locally installed packages
```
Comment (vagimeli) [inline CONTRIBUTING.md:81]:
```suggestion
##### Building locally installed packages 
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -98,6 +100,14 @@ Follow these steps to set up your local copy of the repository:
    bundle install
    ```
 
+##### Building using containerization
```
Comment (vagimeli) [inline CONTRIBUTING.md:103]:
```suggestion
##### Building by using containerization
```

File: _analyzers/index.md [hunk 6]
```diff
@@ -170,6 +170,28 @@ The response provides information about the analyzers for each field:
 }
 ```
 
+## Normalizers
```
Comment (vagimeli) [inline _analyzers/index.md:173]:
```suggestion
## Normalizers

```

File: _analyzers/index.md [hunk 7]
```diff
@@ -170,6 +170,29 @@ The response provides information about the analyzers for each field:
...
+
+Tokenization divides text into individual terms, but it does not address variations in token forms. Normalization resolves these issues by converting tokens into a standard format. This ensures that similar terms are matched appropriately, even if they are not identical.
+
+### Normalization techniques
+
+The following normalization techniques can help address variations in token forms:
```
Comment (vagimeli) [inline _analyzers/index.md:179]:
```suggestion
The following normalization techniques can help address variations in token forms:

```

File: _analyzers/index.md [hunk 8]
```diff
@@ -170,6 +170,30 @@ The response provides information about the analyzers for each field:
...
+
+1. **Case normalization**: Converts all tokens to lowercase to ensure case-insensitive matching. For example, "Hello" is normalized to "hello".
+
+2. **Stemming**: Reduces words to their root form. For instance, "cars" is stemmed to "car", and "running" is normalized to "run".
```
Comment (vagimeli) [inline _analyzers/index.md:183]:
```suggestion
2. **Stemming**: Reduces words to their root form. For instance, "cars" is stemmed to "car" and "running" is normalized to "run".
```

**Comment by vagimeli** [comment]
> > @sb2k16 , @vagimeli , This is a new feature for Data Prepper 2.10, which has not yet been released. We should target this to a new branch. I just created one named `data-prepper-2.10`.
> > @sb2k16 , Can you edit your PR to target the `data-prepper-2.10` branch?
> 
> I have edited the PR to target `data-prepper-2.10` branch.

@dlvenable Once you approve, I'll move this PR forward for publishing.

**Review by dlvenable — APPROVED** [review]
Thank you @sb2k16 !

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+`stream_name` | Yes      | String | The name of each Kinesis stream.
+`initial_position` | No       | String | The position from where `kinesis` source starts reading stream records. `LATEST` starts reading from the most recent stream record. `EARLIEST` starts reading from the begining of the stream. Default is `LATEST`.
+`checkpoint_interval` | No       | Duration | Periodically checkpoint Kinesis streams to avoid duplication of record processing. Default is `PT2M`.
+`compression` | No | String  | Specifies the compression format. To decompress records added by [CloudWatch subscription filter](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/SubscriptionFilters.html) to Kinesis, use the `gzip` compression format.
+
+## codec
+
+The `codec` determines how the `kinesis` source parses each Amazon Kinesis Record. For increased and more efficient performance, you can use [codec combinations]({{site.url}}{{site.baseurl}}/data-prepper/common-use-cases/codec-processor-combinations/) with certain processors.
+
+### `newline` codec
```
Comment (mmehrten) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:63]: Can we add JSON codec here?

**Review by vagimeli — APPROVED** [review]

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+      ],
+      "Resource": [
+        "arn:aws:dynamodb:us-east-1:{account-id}:table/kinesis-pipeline"
+      ]
+    }
+  ]
+}
+```
+
+The `kinesis` source in Data Prepper uses a DynamoDB table for ingestion coordination among multiple workers, so you need DynamoDB permissions.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:146]:
```suggestion
The `kinesis` source uses a DynamoDB table for ingestion coordination among multiple workers, so you need DynamoDB permissions.
```

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+## Configuration options
+
+The `kinesis` source supports the following configuration options.
+
+Option | Required | Type     | Description
+:--- |:---------|:---------| :---
+`aws` | Yes      | AWS      | Specifies the AWS configuration. See [`aws`](#aws).
+`acknowledgments` | No       | Boolean  | When set to `true`, enables the `kinesis` source to receive [end-to-end acknowledgments]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/pipelines#end-to-end-acknowledgments) when events are received by OpenSearch sinks.
+`streams` | Yes      | List     | Configures a list of multiple Kinesis data streams that the Data Prepper `kinesis` source uses to read records. You can configure up to four streams. See [Streams](#streams).
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:40]:
```suggestion
`streams` | Yes      | List     | Configures a list of multiple Kinesis data streams that the `kinesis` source uses to read records. You can configure up to four streams. See [Streams](#streams).
```

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+The `kinesis` source in Data Prepper uses a DynamoDB table for ingestion coordination among multiple workers, so you need DynamoDB permissions.
+
+## Metrics
+
+The `kinesis` source includes the following metrics.
+
+### Counters
+
+* `recordsProcessed`: Counts the number of stream records processed from Kinesis streams.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:154]:
```suggestion
* `recordsProcessed`: Counts the number of processed stream records.
```

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+* `recordProcessingErrors`: Counts the number of processing errors for stream records from Kinesis streams.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:155]:
```suggestion
* `recordProcessingErrors`: Counts the number of stream record processing errors.
```

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+* `acknowledgementSetSuccesses`: Tracks the total number stream records processed that were successfully added to sink.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:156]:
```suggestion
* `acknowledgementSetSuccesses`: Counts the number of processed stream records that were successfully added to the sink. 
```

File: _data-prepper/pipelines/configuration/sources/kinesis.md [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+## Metrics
+
+The `kinesis` source includes the following metrics.
+
+### Counters
+
+* `recordsProcessed`: Counts the number of stream records processed from Kinesis streams.
+* `recordProcessingErrors`: Counts the number of processing errors for stream records from Kinesis streams.
+* `acknowledgementSetSuccesses`: Tracks the total number stream records processed that were successfully added to sink. 
+* `acknowledgementSetFailures`: Tracks the total number stream records processed that failed to be added to sink.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/kinesis.md:157]:
```suggestion
* `acknowledgementSetFailures`: Counts the number of processed stream records that failed to be added to the sink.
```

**Review by natebower — COMMENTED** [review]
@sb2k16 @vagimeli Please see my changes and let me know if you have any questions. Thanks!

**Review by dlvenable — APPROVED** [review]

**Pull request merged**